core/src/main/java/org/apache/calcite/rel/rules/SemiJoinRule.java

```diff
@@ -42,14 +43,18 @@
 /**
  * Planner rule that creates a {@code SemiJoin} from a
  * {@link org.apache.calcite.rel.core.Join} on top of a
- * {@link org.apache.calcite.rel.logical.LogicalAggregate}.
+ * {@link org.apache.calcite.rel.logical.LogicalAggregate} or
+ * on a {@link org.apache.calcite.rel.RelNode} which is
+ * unique for join's right keys.
  */
 public abstract class SemiJoinRule
     extends RelRule<SemiJoinRule.Config>
     implements TransformationRule {
   private static boolean isJoinTypeSupported(Join join) {
     final JoinRelType type = join.getJoinType();
-    return type == JoinRelType.INNER || type == JoinRelType.LEFT;
+    return type == JoinRelType.INNER
+        || type == JoinRelType.LEFT
+        || type == JoinRelType.SEMI;
   }
```

Comment:
If it is already a semi join, do we still need to convert it?

Reply:
Not sure about this... can we have a SEMI with an aggregate inside in a real scenario?

The code that you mention seems to be "legacy code" that we forgot to clean-up when CALCITE-4623 was applied. In practice, the existing `JoinToSemiJoinRule` should not match SEMI because it is not included on its `isJoinTypeSupported` method (which was changed via CALCITE-4623 to consider just INNER and LEFT):
```
  private static boolean isJoinTypeSupported(Join join) {
    final JoinRelType type = join.getJoinType();
    return type == JoinRelType.INNER || type == JoinRelType.LEFT;
  }
```

Reply:
It's good to know there is already a discussion about this before in CALCITE-4623.

Consider this case, we have a plan below:
```xml
LogicalJoin(condition=[=($7, $9)], joinType=[inner])
  LogicalTableScan(table=[[CATALOG, SALES, EMP]])
  LogicalAggregate(group=[{0}], agg#0=[MIN($1)])
    LogicalProject(DEPTNO=[$0], $f0=[true])
      LogicalTableScan(table=[[CATALOG, SALES, DEPT]])
```
After `JoinOnUniqueToSemiJoinRule`, it would be transformed to:
```xml
LogicalJoin(condition=[=($7, $9)], joinType=[semi])
  LogicalTableScan(table=[[CATALOG, SALES, EMP]])
  LogicalAggregate(group=[{0}], agg#0=[MIN($1)])
    LogicalProject(DEPTNO=[$0], $f0=[true])
      LogicalTableScan(table=[[CATALOG, SALES, DEPT]])
```
And we can still use `JoinToSemiJoinRule` to remove the `Aggregate`:
```xml
LogicalJoin(condition=[=($7, $9)], joinType=[semi])
  LogicalTableScan(table=[[CATALOG, SALES, EMP]])
  LogicalProject(DEPTNO=[$0], $f0=[true])
    LogicalTableScan(table=[[CATALOG, SALES, DEPT]])
```

In many cases, we would not expect the rules firing orders, especially in `HepPlanner`, hence the above transforming order is valid and should be valid IMHO.

Reply:
It seems a bit weird to do this optimization in a rule named JoinToSemiJoinRule. Maybe it's better to move this optimization to another place.

Reply:
@chunweilei I figured out another way to do this. In `JoinOnUniqueToSemiJoinRule`, if we found that right is `Aggregate`, we can fallback to `ProjectToSemiJoinRule`.